### PR TITLE
Laboratory work #1, Sofya Kostina - 21FPL2

### PR DESCRIPTION
**Pull request by d07l**

### CODE REVIEW

File: lab_1_keywords_tfidf/main.py [hunk 1]
```diff
@@ -17,6 +17,15 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
...
     """
+    if not isinstance(text, str):
+        return None
+    text = text.replace('\n', ' ').lower().strip()
+    fin_text = ''
+    for symbol in text:
+        if symbol.isalnum() == True or symbol == ' ':
+            fin_text += symbol
+    tokens = fin_text.split()
+    return tokens
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:28]: remove `paas`

File: lab_1_keywords_tfidf/main.py [hunk 2]
```diff
@@ -33,6 +42,14 @@ def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(tokens, list) or not isinstance(stop_words, list):
+        return None
+    for i in tokens:
+        for j in stop_words:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:48]: I believe your implementation works, but let's try to make it more pythonic - for example, use the `i in stop_words` construct

File: lab_1_keywords_tfidf/main.py [hunk 3]
```diff
@@ -33,6 +42,14 @@ def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list
...
+            if i == j:
+                while j in tokens:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:50]: loop inside the loop inside the loop - too complicated

File: seminars/practice_1_intro.py [hunk 4]
```diff
@@ -28,7 +28,7 @@
 # third_num = first_num / second_num
 
 # Fix the first exception:
-second_num = 6
+second_num = 5
```
Comment (artyomtugaryov) [inline seminars/practice_1_intro.py:31]: revert the changes

**Review by artyomtugaryov — COMMENTED** [review]
Not bad, but let's try pythonize the code

File: lab_1_keywords_tfidf/main.py (hunk 2 above)
Comment (d07l) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 3 above)
Comment (d07l) [reply lab_1_keywords_tfidf/main.py]: fixed

File: seminars/practice_1_intro.py (hunk 4 above)
Comment (d07l) [reply seminars/practice_1_intro.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 1 above)
Comment (d07l) [reply lab_1_keywords_tfidf/main.py:29]: done

File: lab_1_keywords_tfidf/main.py [hunk 5]
```diff
@@ -17,7 +18,15 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not isinstance(text, str):
+        return None
+    text = text.replace('\n', ' ').lower().strip()
+    fin_text = ''
+    for symbol in text:
+        if symbol.isalnum() is True or symbol == ' ':
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:26]: simplify the condition. Don't use `== True`

File: lab_1_keywords_tfidf/main.py [hunk 6]
```diff
@@ -48,7 +65,17 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
...
     """
-    pass
+    if not isinstance(tokens, list):
+        return None
+
+    frequencies = {}
+    for i in tokens:
+        if not isinstance(i, str):
+            return None
+        freq = tokens.count(i)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:75]: Why do you need `freq`? Can you set `tokens.count(i)` directly to `frequencies[i]`?

File: lab_1_keywords_tfidf/main.py [hunk 7]
```diff
@@ -82,7 +119,22 @@ def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
...
+    term_freq = {}
+    length = 0
+
+    for k, val in frequencies.items():
+        if not isinstance(k, str) or not isinstance(val, int):
+            return None
+        length = length + val
+
+    for k, val in frequencies.items():
+        term_f = val / length
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:134]: again

File: lab_1_keywords_tfidf/main.py [hunk 8]
```diff
@@ -99,7 +151,25 @@ def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optio
...
-    pass
+    if not isinstance(term_freq, dict) or not isinstance(idf, dict):
+        return None
+    if term_freq == {}:
+        return None
+
+    for k, val in term_freq.items():
+        if not isinstance(k, str) or not isinstance(val, float):
+            return None
+        if idf == {}:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:162]: bad condition. an empty dict equals `False` - this is a Falsy object

File: lab_1_keywords_tfidf/main.py [hunk 9]
```diff
@@ -99,7 +151,25 @@ def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optio
...
+            term_freq[k] = val * math.log(47 / 1)
+        else:
+            for key1, val1 in idf.items():
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:165]: You don't need this for - use `idf[key1]`

File: lab_1_keywords_tfidf/main.py (hunk 5 above)
Comment (d07l) [reply lab_1_keywords_tfidf/main.py]: fixed

File: lab_1_keywords_tfidf/main.py (hunk 6 above)
Comment (d07l) [reply lab_1_keywords_tfidf/main.py]: fixed

File: lab_1_keywords_tfidf/main.py (hunk 7 above)
Comment (d07l) [reply lab_1_keywords_tfidf/main.py]: fixed

File: lab_1_keywords_tfidf/main.py (hunk 8 above)
Comment (d07l) [reply lab_1_keywords_tfidf/main.py]: fixed

File: lab_1_keywords_tfidf/main.py (hunk 9 above)
Comment (d07l) [reply lab_1_keywords_tfidf/main.py]: fixed

**Review by artyomtugaryov — APPROVED** [review]

**Comment by artyomtugaryov** [comment]
Remove the `1` file - `git rm lab_1_keywords_tfidf/1` -> git commit
